src/System.Runtime.Extensions/src/System/Environment.Win32.cs

```diff
@@ -208,14 +208,185 @@ private unsafe static char[] GetEnvironmentCharArray()
+                    break;
+                case SpecialFolder.CommonStartup:
+                    folderGuid = Interop.mincore.KnownFolders.CommonStartup;
+                    break;
+                case SpecialFolder.CommonDesktopDirectory:
+                    folderGuid = Interop.mincore.KnownFolders.PublicDesktop;
+                    break;
+                case SpecialFolder.CommonTemplates:
+                    folderGuid = Interop.mincore.KnownFolders.CommonTemplates;
+                    break;
+                case SpecialFolder.CommonVideos:
+                    folderGuid = Interop.mincore.KnownFolders.PublicVideos;
+                    break;
+                case SpecialFolder.Fonts:
+                    folderGuid = Interop.mincore.KnownFolders.Fonts;
+                    break;
+                case SpecialFolder.NetworkShortcuts:
+                    folderGuid = Interop.mincore.KnownFolders.NetHood;
+                    break;
+                case SpecialFolder.PrinterShortcuts:
+                    folderGuid = Interop.mincore.KnownFolders.PrintersFolder;
+                    break;
+                case SpecialFolder.UserProfile:
+                    folderGuid = Interop.mincore.KnownFolders.Profile;
+                    break;
+                case SpecialFolder.CommonProgramFilesX86:
+                    folderGuid = Interop.mincore.KnownFolders.ProgramFilesCommonX86;
+                    break;
+                case SpecialFolder.ProgramFilesX86:
+                    folderGuid = Interop.mincore.KnownFolders.ProgramFilesX86;
+                    break;
+                case SpecialFolder.Resources:
+                    folderGuid = Interop.mincore.KnownFolders.ResourceDir;
+                    break;
+                case SpecialFolder.LocalizedResources:
+                    folderGuid = Interop.mincore.KnownFolders.LocalizedResourcesDir;
+                    break;
+                case SpecialFolder.SystemX86:
+                    folderGuid = Interop.mincore.KnownFolders.SystemX86;
+                    break;
+                case SpecialFolder.Windows:
+                    folderGuid = Interop.mincore.KnownFolders.Windows;
+                    break;
                 default:
-                    // TODO: SHGetFolderPath is not available in the approved API list
                     return string.Empty;
             }
+
+            return GetKnownFolderPath(folderGuid, option);
+        }
+
+        private static string GetKnownFolderPath(string folderGuid, SpecialFolderOption option)
+        {
+            Guid folderId = new Guid(folderGuid);
+
+            string path;
+            int hr = Interop.mincore.SHGetKnownFolderPath(folderId, (uint)option, IntPtr.Zero, out path);
+            if (hr != 0) // Not S_OK
+            {
+                if (hr == Interop.mincore.COR_E_PLATFORMNOTSUPPORTED)
+                {
+                    throw new PlatformNotSupportedException();
```

Comment:
Why do we throw here rather than just returning string.Empty?  In what situation will this return PLATFORMNOTSUPPORTED?